corehq/motech/repeaters/models.py

```diff
@@ -773,6 +781,17 @@ def cancel(self):
         self.next_check = None
         self.cancelled = True
 
+    def can_enqueue(self):
```

Comment:
This `can_enqueue` method makes me a little uncomfortable because it changes state even though it looks like a benign state check. If you call it twice quickly the second result will always be false.

What do you think of renaming it to `forward_now` and moving `process_repeat_record.delay(repeat_record)` to where it returns true?

```py
def forward_now(self):
    from corehq.motech.repeaters.tasks import process_repeat_record
    if self.next_check > datetime.utcnow():
        # self is already in the queue, or it is not ready to be enqueued
        return
    # Set the next check to happen an arbitrarily long time from now so
    # if something goes horribly wrong with the delayed task it will not
    # be lost forever. This check is expected to occur rarely, if ever,
    # because `process_repeat_record` will either succeed or will reset
    # the next check to something sooner.
    self.next_check = datetime.utcnow() + timedelta(hours=48)
    try:
        self.save()
    except ResourceConflict:
        # Another process beat us to the punch. This takes advantage of
        # the fact Couch uses optimistic locking to prevent one process
        # from overwriting the work of another with stale data.
        return
    process_repeat_record.delay(self)
```

As you can see I added some verbose comments so future readers can get a better handle on what's happening here. I also made another somewhat critical change to prevent a very old `next_check` value from being incremented to a time that is not two days from now:
```diff
- self.next_check += timedelta(hours=48)
+ self.next_check = datetime.utcnow() + timedelta(hours=48)`
```
The places that call `record.can_enqueue()` will need to be changed to call `record.forward_now()`

Reply:
Maybe it should be named `attempt_forward_now`

Reply:
Thank you. That not only addresses the issue of the state change, but also resolves something I was uncomfortable about; separating that state change from the purpose of the state change. It felt wrong to me that we were changing `next_check` to indicate that the record was queued separately from actually putting it in the repeat record queue. I'm glad this resolves both. I'll make the change.